src/library/scala/collection/immutable/VectorMap.scala

```diff
@@ -0,0 +1,155 @@
+  *
+  *  Unlike `ListMap`, `VectorMap` has amortized effectively constant lookup at the expense
+  *  of using extra memory and generally lower performance for other operations
+  *
+  *  @tparam K      the type of the keys contained in this vector map.
+  *  @tparam V      the type of the values associated with the keys in this vector map.
+  *
+  * @author Matthew de Detrich
+  * @version 2.13
+  * @since 2.13
+  * @define coll immutable vector map
+  * @define Coll `immutable.VectorMap`
+  */
+final class VectorMap[K, +V] private[immutable] (
+    val fields: Vector[K],
+    val underlying: Map[K, (Int, V)])
+    extends AbstractMap[K, V]
+    with LinkedMap[K, V]
+    with MapOps[K, V, VectorMap, VectorMap[K, V]]
+    with StrictOptimizedIterableOps[(K, V), Iterable, VectorMap[K, V]] {
+
+  override protected[this] def className: String = "VectorMap"
+
+  override def updated[V1 >: V](key: K, value: V1): VectorMap[K, V1] = {
+    if (underlying.get(key).isDefined) {
+      val oldKey = underlying(key)._1
+      new VectorMap(fields,
+                    underlying.updated(key, (oldKey, value.asInstanceOf[V])))
+    } else {
+      new VectorMap(
+        fields :+ key,
+        underlying.updated(key, (fields.length + 1, value.asInstanceOf[V])))
+    }
+  }
+
+  override def withDefault[V1 >: V](d: K => V1): Map.WithDefault[K, V1] =
+    new Map.WithDefault(this, d)
+
+  override def withDefaultValue[V1 >: V](d: V1): Map.WithDefault[K, V1] = new Map.WithDefault[K, V1](this, _ => d)
+
+  override def iterator: Iterator[(K, V)] = new Iterator[(K, V)] {
+    private val fieldsIterator = fields.iterator
+
+    override def hasNext: Boolean = fieldsIterator.hasNext
+
+    override def next(): (K, V) = {
+      val field = fieldsIterator.next()
+      (field, underlying(field)._2)
+    }
+  }
+
+  override def get(key: K): Option[V] = underlying.get(key) match {
+    case Some(v) => Some(v._2)
+    case None => None
+  }
+
+  override def remove(key: K): VectorMap[K, V] = {
+    underlying.get(key) match {
+      case Some((index, _)) =>
+        new VectorMap(fields.patch(index, Nil, 1), underlying - key)
+      case _ =>
+        this
+    }
+  }
+
+  override def mapFactory: MapFactory[VectorMap] = VectorMap
+
+  override def size: Int = fields.size
+
+  override def knownSize: Int = fields.size
+
+  override def isEmpty: Boolean = fields.isEmpty
+
+  override final def contains(key: K): Boolean = underlying.contains(key)
+
+  override def head: (K, V) = iterator.next()
+
+  override def last: (K, V) = {
+    val last = fields.last
+    (last, underlying(last)._2)
+  }
+
+  override def lastOption: Option[(K, V)] = {
+    fields.lastOption match {
+      case Some(last) => Some(last, underlying(last)._2)
+      case None => None
+    }
+  }
+
+  override def tail: VectorMap[K, V] = {
+    val tail = fields.last
+
+    new VectorMap(fields.slice(1, fields.size), underlying.remove(tail))
```

Comment:
`fields.tail` is probably both more readable and more efficient than using `.slice`.